File: src/data/codes/adaboost.m
clc;  % AdaBoost
clear;
close all;

%% =========================
% SETTINGS
% ==========================
useSynthetic = true;
csvFile = 'adaboost_driver_risk.csv';
numLearners = 50;

%% =========================
% LOAD OR CREATE DATA
% ==========================
if useSynthetic
    rng(21);
    n = 180;

    Avg_Speed = randi([60 140], n, 1);
    Harsh_Brakes = randi([0 20], n, 1);
    Night_Driving_Hours = randi([0 25], n, 1);
    Mobile_Usage = randi([0 60], n, 1);
    Accident_History = randi([0 4], n, 1);

    riskScore = 0.03*Avg_Speed + 0.18*Harsh_Brakes + 0.12*Night_Driving_Hours ...
                + 0.08*Mobile_Usage + 0.9*Accident_History;

    Risk_Level = double(riskScore > median(riskScore));

    data = table(Avg_Speed, Harsh_Brakes, Night_Driving_Hours, ...
                 Mobile_Usage, Accident_History, Risk_Level);
else
    data = readtable(csvFile);
end

disp(head(data));

%% PREPROCESSING
data = rmmissing(data);
data = unique(data);

X = [data.Avg_Speed, data.Harsh_Brakes, data.Night_Driving_Hours, ...
     data.Mobile_Usage, data.Accident_History];
Y = data.Risk_Level;

%% TRAIN-TEST SPLIT
cv = cvpartition(Y, 'HoldOut', 0.2);
X_train = X(training(cv), :);
Y_train = Y(training(cv), :);
X_test  = X(test(cv), :);
Y_test  = Y(test(cv), :);

%% MODEL BUILDING
model = fitcensemble(X_train, Y_train, ...
    'Method', 'AdaBoostM1', ...
    'NumLearningCycles', numLearners);

%% PREDICTION
Y_pred = predict(model, X_test);

%% EVALUATION
confMat = confusionmat(Y_test, Y_pred);
accuracy = sum(diag(confMat)) / sum(confMat(:));
TP = confMat(2,2); FP = confMat(1,2); FN = confMat(2,1);
precision = TP/(TP+FP+eps);
recall = TP/(TP+FN+eps);
f1 = 2*precision*recall/(precision+recall+eps);

disp('Confusion Matrix:');
disp(confMat);
fprintf('\nAccuracy  = %.4f\n', accuracy);
fprintf('Precision = %.4f\n', precision);
fprintf('Recall    = %.4f\n', recall);
fprintf('F1-score  = %.4f\n', f1);

figure;
confusionchart(Y_test, Y_pred);
title('AdaBoost Confusion Matrix');

File: src/data/codes/adaboost.py
"""AdaBoost (SAMME = discrete AdaBoost multiclass)
# pip install numpy pandas scikit-learn matplotlib
"""
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score
from sklearn.metrics import ConfusionMatrixDisplay
import matplotlib.pyplot as plt

USE_SYNTHETIC = True
N_ESTIMATORS = 50
RNG = np.random.default_rng(21)

if USE_SYNTHETIC:
    n = 180
    speed = RNG.integers(60, 141, size=n)
    brakes = RNG.integers(0, 21, size=n)
    night = RNG.integers(0, 26, size=n)
    mobile = RNG.integers(0, 61, size=n)
    acc_hist = RNG.integers(0, 5, size=n)
    risk_score = (
        0.03 * speed
        + 0.18 * brakes
        + 0.12 * night
        + 0.08 * mobile
        + 0.9 * acc_hist
    )
    risk_level = (risk_score > np.median(risk_score)).astype(int)
    data = pd.DataFrame(
        {
            "Avg_Speed": speed,
            "Harsh_Brakes": brakes,
            "Night_Driving_Hours": night,
            "Mobile_Usage": mobile,
            "Accident_History": acc_hist,
            "Risk_Level": risk_level,
        }
    )
else:
    data = pd.read_csv("adaboost_driver_risk.csv")

print(data.head())
data = data.dropna().drop_duplicates()
cols = ["Avg_Speed", "Harsh_Brakes", "Night_Driving_Hours", "Mobile_Usage", "Accident_History"]
X = data[cols].values
y = data["Risk_Level"].values

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, stratify=y, random_state=42
)

base = DecisionTreeClassifier(max_depth=1, random_state=42)
model = AdaBoostClassifier(
    estimator=base, n_estimators=N_ESTIMATORS, random_state=42
).fit(X_train, y_train)
y_pred = model.predict(X_test)

print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))
print(f"Accuracy  = {accuracy_score(y_test, y_pred):.4f}")
print(f"Precision = {precision_score(y_test, y_pred, zero_division=0):.4f}")
print(f"Recall    = {recall_score(y_test, y_pred, zero_division=0):.4f}")
print(f"F1-score  = {f1_score(y_test, y_pred, zero_division=0):.4f}")

fig, ax = plt.subplots()
ConfusionMatrixDisplay.from_predictions(y_test, y_pred, ax=ax)
ax.set_title("AdaBoost Confusion Matrix")
plt.show()
